### DISCUSSION TITLE
Feedback + questions

### DISCUSSION THREAD

**Discussion by brillout** [hattipjs/hattip#58] [General]
By @bholmesdev:

> Also for Hattip: really like that it’s as simple as request / response. The compose API for middleware is also excellent. Trouble I ran into: I couldn’t get traditional http-server middleware working at all. A ton of middleware depends on passing a response in using writeHead, which isn’t part of the web standard Response spec

> Curious if you have suggestions here? Also, if you have suggestions for a static assets server so I don’t have to write one myself?

**Comment by brillout** [comment]
> I couldn’t get traditional http-server middleware working at all

Do you mean https://www.npmjs.com/package/http-server or Node.js's `http` module?

> A ton of middleware depends on passing a response in using writeHead, which isn’t part of the web standard Response spec

Yes, I don't think there is an (easy) way to port these to the new Request/Response standard. If you want to use these middlewares, then you'll have to use HatTip's `@hattip/adapter-node`.

> if you have suggestions for a static assets server

IIRC @cyco130 was (considering) working on this.

**Reply by bholmesdev** [reply]
> I couldn’t get traditional http-server middleware working at all

Yes, I'm referring to the `http` module. I was specifically wiring up the [server-handler package](https://www.npmjs.com/package/serve-handler)

> A ton of middleware depends on passing a response in using writeHead, which isn’t part of the web standard Response spec

Gotcha, I'm fine using the node adapter! What's the config option to toggle this?

**Reply by brillout** [reply]
> What's the config option to toggle this?

The readme has an example of doing this.

I'm actually realizing that it's an interesting use case: to use HatTip to write modern Request/Response server code while using Node.js(/Express.js) middlewares. You obviously lose the "deploy to anywhere/edge" part and we should clearly communicate the trade off.

Created a ticket for adding a static middleware: https://github.com/hattipjs/hattip/issues/59.

**Comment by cyco130** [comment]
Hi @bholmesdev,

Thanks for the feedback.

This is how I use Node middlewares (`req, res, next` style)  with HatTip:

```js
// entry-node.js
import { createServer } from 'node:http';
import sirv from 'sirv';
import { createMiddleware } from '@hattip/adapter-node';
import hattipHandler from './entry-hattip.js';

const sirvMiddleware = sirv('public');
const hattipMiddleware = createMiddleware(hattipHandler);

createServer((req, res) =>
  sirvMiddleware(req, res, () => hattipMiddleware(req, res))
).listen(3000, () => {
  console.log(`Server listening on http://localhost:3000`);
});
```

Here's a [live example](https://stackblitz.com/edit/node-old2rq?file=package.json,entry-node.js,entry-hattip.js).

The idea is that your HatTip app (which lives in `entry-hattip.js` in this example) becomes a middleware itself and then you chain it with whatever other middleware you want. Here I did the chaining manually (`sirvMiddleware(req, res, () => hattipMiddleware(req, res))`) to avoid adding a dependency but you can use Express, Connect, or [Polka](https://github.com/lukeed/polka) if you have more sophisticated middleware chaining needs, e.g. mounting middleware to subpaths.

On static file serving middleware, it's on the map but I didn't prioritize it because all target platforms have their on semi-automatic way to handle static files anyway. Node doesn't, but I used to use HatTip almost exclusively in combination with [vavite](https://github.com/cyco130/vavite) which handles it for me. Now [HatTip has its own CLI and Vite plugin](https://github.com/hattipjs/hattip/tree/main/packages/cli/vite) which makes it even easier.

In addition to the issue @brillout has opened, we've had some discussion on how to proceed: #45. Now that we have a form data parser, cross-platform filesystem handling (where it makes sense) is the next logical step. Even on platforms that don't have a filesystem, having something backed by other storage media like memory, S3, KV stores, Redis etc might make sense. A static file server would be then more or less trivial to implement.

**Reply by bholmesdev** [reply]
Ah, _super_ helpful example on integrating with `node:http`. This is honestly what I was looking for in the `@hattip/adapter-node` documentation but only saw the express example 👍

**Reply by cyco130** [reply]
Took your word and updated the readme.
